### PR TITLE
#20 Labelled Exits

### PR DESCRIPTION
**Pull request by bilal-chughtai**
Implemented the rule outlined in #20 , along with some tests for it. One could also check for labels on do statements, though this is redundant, as code will not compile if an exit statement label exists without a matching do statement label. 

Some background on the reason for wanting this via @AdamVoysey: The behaviour of unlabeled exit statements has subtly changed between different versions of the Fortran standard as new functionality has been added. So rather than say something like "if you're using Fortran 2003 you have to label these cases; if you are using Fortran 2008 you also have to label these" etc. we decided it was just easier to label everything, as that is not ambiguous or dependant on which compiler settings you use.

### CODE REVIEW

File: unit-tests/fortran_labelled_exit_test.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+"""
+from typing import List
+
+import pytest  # type: ignore
+# ToDo: Obviously we shouldn't be importing "private" modules but until pytest
+#       sorts out its type hinting we are stuck with it.
+#
+from _pytest.fixtures import FixtureRequest  # type: ignore
+
+import stylist.fortran
```
Comment (MatthewHambley) [inline unit-tests/fortran_labelled_exit_test.py:18]: Lets be specific about what we are importing.

File: source/stylist/fortran.py [hunk 2]
```diff
@@ -200,6 +203,34 @@ def examine_fortran(self, subject: FortranSource) -> List[Issue]:
...
+    """
+
+    def examine_fortran(self, subject: FortranSource) -> List[Issue]:
+        issues = []
+
+        for exit in subject.find_all(Fortran2003.Exit_Stmt):
+            if exit.items[1] is None:
+                issues.append(Issue('Usage of "exit" without label '
+                                    'indicating '
+                                    'which do construct is being exited from.',
```
Comment (MatthewHambley) [inline source/stylist/fortran.py:218]: Throughout change unquoted references to "do" construct are hard to read. They don't scan well. This is particularly noticeable in output strings such as this and documentation strings.

File: unit-tests/fortran_labelled_exit_test.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        expectation: List[str] = []
+        message = '{line}: Usage of "exit" without label indicating which ' \
+                  'do construct is being exited from.'
+        if do_construct_name == '':
+            expectation.extend([
+                message.format(line=6),
+                message.format(line=12)
+            ])
+        expectation.sort(key=lambda x: (int(x.split(':', 1)[0]),
+                                        x.split(':', 1)))
```
Comment (MatthewHambley) [inline unit-tests/fortran_labelled_exit_test.py:66]: This seems to sort the entries by line number. It is messy and complicated but is it necessary? The items should appear in the list in the order they were added, in which case they should be correctly ordered already.

File: unit-tests/fortran_labelled_exit_test.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        text = template.format(
+            do_construct_name=do_construct_name)
+        print(text)  # Shows up in failure reports, for debugging
+        reader = SourceStringReader(text)
+        source = FortranSource(reader)
+        unit_under_test = stylist.fortran.LabelledExit()
+        issues = unit_under_test.examine(source)
+        issue_descriptions = [str(issue) for issue in issues]
+        issue_descriptions.sort(key=lambda x: (int(x.split(':', 1)[0]),
+                                               x.split(':', 1)))
```
Comment (MatthewHambley) [inline unit-tests/fortran_labelled_exit_test.py:77]: Is this sort necessary? Lines should be returned in order from the rule.

**Review by MatthewHambley — CHANGES_REQUESTED** [review]
Change looks good. Please respond to comments with changes or explinations.

File: unit-tests/fortran_labelled_exit_test.py (hunk 1 above)
Comment (bilal-chughtai) [reply unit-tests/fortran_labelled_exit_test.py:18]: fixed.

File: source/stylist/fortran.py (hunk 2 above)
Comment (bilal-chughtai) [reply source/stylist/fortran.py:218]: Agreed - All now read "do"

File: unit-tests/fortran_labelled_exit_test.py (hunk 3 above)
Comment (bilal-chughtai) [reply unit-tests/fortran_labelled_exit_test.py:66]: It's not - it was a remnant from copying over and modifying another unit test. Removed.

File: unit-tests/fortran_labelled_exit_test.py (hunk 4 above)
Comment (bilal-chughtai) [reply unit-tests/fortran_labelled_exit_test.py:77]: Again, no. Removed.

**Review by MatthewHambley — APPROVED** [review]
Everything looks fine for a a final merge.

**Pull request merged**